plasma_framework/contracts/src/fast_exits/liquidity.sol

```diff
@@ -0,0 +1,202 @@
+    */
+    function startExitOnRootchainContract(
+        uint256 _utxoPosToExit,
+        bytes memory _rlpOutputTxToContract,
+        bytes memory _outputTxToContractInclusionProof,
+        bytes memory _rlpInputCreationTx,
+        bytes memory _inputCreationTxInclusionProof,
+        uint256 _utxoPosInput
+    ) public payable {
+
+        verifyOwnership(_rlpInputCreationTx, _utxoPosInput);
+
+
+            PaymentTransactionModel.Transaction memory decodedSecondTx
+         = PaymentTransactionModel.decode(_rlpOutputTxToContract);
+        require(
+            decodedSecondTx.inputs[0] == bytes32(_utxoPosInput),
+            "Wrong utxoPosInput provided"
+        );
+
+        verifyTxValidity(
+            _utxoPosInput,
+            _rlpInputCreationTx,
+            _inputCreationTxInclusionProof
+        );
+        runExit(
+            _utxoPosToExit,
+            _rlpOutputTxToContract,
+            _outputTxToContractInclusionProof
+        );
+
+        // store the resultant exitid as a trait for the nft and map it to the msg.sender
+        uint160 _exitId = getExitId(_rlpOutputTxToContract, 0);
+        userExitIds[msg.sender].push(_exitId);
+        exitIdtoUser[_exitId] = msg.sender;
+
+        // associate the amount exiting to the exitId
+
+            FungibleTokenOutputModel.Output memory outputFromSecondTransaction
+         = decodedSecondTx.outputs[0];
+        uint256 amount = outputFromSecondTransaction.amount;
+        exitIdtoAmount[_exitId] = amount;
+    }
+
+    /**
+     * @notice Check if the person calling is the same person who created the tx to the contract
+     * @param _rlpInputCreationTx RLP-encoded first transaction that transfers to this contract
+     * @param _utxoPosInput position of the output that created the inputs for second transaction
+    */
+    function verifyOwnership(
+        bytes memory _rlpInputCreationTx,
+        uint256 _utxoPosInput
+    ) internal {
+
+            PaymentTransactionModel.Transaction memory decodedFirstTx
+         = PaymentTransactionModel.decode(_rlpInputCreationTx);
+        uint16 firstTransactionOutputIndex = getOutputIndex(_utxoPosInput);
+
+            FungibleTokenOutputModel.Output memory outputFromFirstTransaction
+         = decodedFirstTx.outputs[firstTransactionOutputIndex];
+        address ownerFirstTxOutput = address(
+            uint160(outputFromFirstTransaction.outputGuard)
+        );
+        require(
+            ownerFirstTxOutput == msg.sender,
+            "First Tx Provided was not called by sender"
+        );
+    }
+
+    /**
+     * @notice Verify the First Tx provided is valid
+     * @param _utxoPosInput position of the output that created the inputs for second transaction
+     * @param _rlpInputCreationTx RLP-encoded first transaction that transfers to this contract
+     * @param _inputCreationTxInclusionProof First transactions inclusion proofs
+    */
+    function verifyTxValidity(
```

Comment:
I think it might not be necessary to check whether the input tx is included to a block or not (?)
If you start exit from a tx that has input not included and succeed, that is an invalid tx in the block, and the child chain is byzantine so all user would mass exit to safe guard their fund.

Reply:
Oh, I think validating the input tx also proves that you are exiting the tx that was meant for you through the LC, since I could provide for a fake tx and could get away starting someone else's exit through the LC ?

Reply:
kinda lost, isn't what you describe done in `verifyOwnership`?
function here is proofing the input tx is indeed included in a block?

Reply:
@boolafish We have to check both these conditions.

Scenario:
1. Alice owns on plasma an `utxo1` - created by `tx1`;
1. Alice transfers the `utxo1` to the liquidity contract (LC)  - `tx2` and `utxo2`;
1. Alice starts a SE via LC.

Let's consider what happens without these assertions.
Since `utxo2` is owned by the LC -> anybody (provided he/she has proofs) can start an exit via LC - as the action on plasma framework must be triggered by LC, which in this case can be triggered by anybody. The contract does not have any additional data, so it assumes that the owner of the funds is the `msg.sender`.

To mitigate this, we assumed that we should check who is the owner of `utxo1` - input to `tx2`. Now, we should provide `tx1` to check whether the owner of `utxo1` was indeed `msg.sender`.

But what if Bob provides `tx1'`, where `utxo1'` is differs from `utxo1` just by the owner - we have to check which tx is the valid - this is why we also check inclusion proof.

---
I was thinking about some other approach - do not make such validations at the beginning, but provide a challenge mechanism. Bob could start the SE from `utxo2` - and initially, he owns it, but Alice can then provide proof that she is the real owner -> the NFT is transferred right to Alice, Bob has lost his bonds.

Reply:
yes yes, and following from what @pik694 said, so `verifyOwnership` checks if you are really the owner of `utxo1` which was an input to `tx2`. And `verifyTxValidity` checks if the `tx1` which you provided to prove the previous statement, is actually valid.

Reply:
And, the challenge mechanism sure is a great idea. But will restricting the problem at the roots be the easier way from a user's perspective?

Reply:
> But will restricting the problem at the roots be the easier way from a user's perspective?

That is right, this is why I created a whole issue ( #627) to discuss which path we should take.

Reply:
oh, thanks for explanation. So the goal is to verify that the `input` is indeed pointing to that specific `tx` provided. Can you rename the function name or try to make the intention more easy to understand?

Reply:
Sure! Any suggestions?

Reply:
probably sth like `verifyTxMatchThePosition` ?